### ISSUE TITLE
模型多卡训练卡死

### ISSUE THREAD

**Issue by Yvette1993** [Peterande/D-FINE#39]
您好，非常感谢您功效的代码。
在多卡训练时出现了卡死，然后报NCCL通信超时的错误，看了RT-DETR中的一个解决方案是vfl的loss为nan时返回0。但是改过后模型依然卡死，进行了bug发现卡死的位置在loss之后的scaler.scale(loss).backward()。请问有什么解决方案吗？下图是在det_engine.py中插入的print:
![JPEG_20241108_111345_2996751310393036089](https://github.com/user-attachments/assets/35788b9a-9e87-4eab-a635-1938e6d5dc50)
和相应的输出：
![JPEG_20241108_111601_2007961292566673349](https://github.com/user-attachments/assets/3d23c499-64ee-438a-90cd-d83d9e7afb9d)
训练用了7张A100,total_batch_size=28,seed=0,num_works=0

**Comment by Peterande** [comment]
DDP训练时其他机器遇到问题0卡会在这个位置等，你输出的只是0卡的结果，并不能看出真实的报错是什么

**Comment by Yvette1993** [comment]
我重新print了0-6卡的信息，每个都是卡在loss输出之后，哭了

**Comment by Yvette1993** [comment]
训练的数据集是自己的，数据集中有图片是没有目标的

**Comment by Peterande** [comment]
单卡训练会报错吗？

**Comment by Yvette1993** [comment]
单卡，total_batch_size=1.报错如下
![Uploading JPEG_20241108_141604_455998059713961990.jpg…]()

**Comment by Yvette1993** [comment]
搜了一下说是有部分loss没有参与单向传播，把DDP中find_unused_parameters=True后，单卡可以正常训练，但是多卡卡在了test5–---之后

**Comment by Peterande** [comment]
似乎报错图片没有上传完成：
![image](https://github.com/user-attachments/assets/45191efb-7bb0-4a00-b758-9b37690ad2e1)

**Comment by Yvette1993** [comment]
单卡报错如下：
![JPEG_20241108_152209_4617550157978139377](https://github.com/user-attachments/assets/9b441351-816c-4046-9ac7-e43a7416a117)
改了DDP参数后，没有报错但是卡在了test5-----之后，如下：
![JPEG_20241108_152404_3001210942678176059](https://github.com/user-attachments/assets/579c3481-6ca6-4e89-95d1-3ee6721709bc)
然后需要手动Ctrl+C了……

**Comment by Peterande** [comment]
> 单卡报错如下： ![JPEG_20241108_152209_4617550157978139377](https://private-user-images.githubusercontent.com/30230189/384265878-9b441351-816c-4046-9ac7-e43a7416a117.jpg?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzEwNTEwMzIsIm5iZiI6MTczMTA1MDczMiwicGF0aCI6Ii8zMDIzMDE4OS8zODQyNjU4NzgtOWI0NDEzNTEtODE2Yy00MDQ2LTlhYzctZTQzYTc0MTZhMTE3LmpwZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDExMDglMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQxMTA4VDA3MjUzMlomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTdhMThlZjNhMmI0NDcxODhhY2RlOGYyZDBhYmIwNTU2Mjk2MzQ0ODRmZTljMzIyZjM4MWE2NTdmN2M4YzljOGQmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.82DZ4sNmxznZYbPb3weiwPp7yOeDHhoTnJGP5SBmrqc) 改了DDP参数后，没有报错但是卡在了test5-----之后，如下： ![JPEG_20241108_152404_3001210942678176059](https://private-user-images.githubusercontent.com/30230189/384266374-579c3481-6ca6-4e89-95d1-3ee6721709bc.jpg?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzEwNTEwMzIsIm5iZiI6MTczMTA1MDczMiwicGF0aCI6Ii8zMDIzMDE4OS8zODQyNjYzNzQtNTc5YzM0ODEtNmNhNi00ZTg5LTk1ZDEtM2VlNjcyMTcwOWJjLmpwZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDExMDglMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQxMTA4VDA3MjUzMlomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTIxMWIxNzVkMzA1MGJhMWJhNWE1MWE1YTkyMjcxNmYxZWUyNTRiN2NhMTYxMzMwMDYzMjZkZmE1Zjk3YTNkMWMmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.CaT10JqyDz_1QZBbskiZz4KOzJvdchxq2Fp3pDo1v-I) 然后需要手动Ctrl+C了……

第一个报错应该是你新加了参数，但是这些参数没有梯度。
如果您做了修改，我可能没法判断问题出在哪，目前没有其他人反应类似问题。

**Comment by Yvette1993** [comment]
没有加新的参数呢，只对数据集地址和类别名和数量进行了调整，还有就是total_batch_size的参数了。现在单卡的训练修改了DDP的find_unsed_parameters=True后可以正常训练了。但是多卡还是不行

**Comment by Peterande** [comment]
还是建议重新clone一下代码再试试。find_unsed_parameters就是说你有没有收到梯度的参数，正常情况下是不会报错的

**Comment by olojuwin** [comment]
我遭遇了同样的问题，单卡可以训练，多卡会卡死

**Comment by Yvette1993** [comment]
重新clone了代码还是同样的问题……

**Comment by Peterande** [comment]
https://github.com/Peterande/D-FINE/issues/19

请提供一下启动命令和所有修改的内容。
同时麻烦截图一下卡死时候的nvitop界面，看一下卡住时候的显存和内存占用。卡死的一种可能是显存/内存溢出，请确保机器上的显存/内存足够。

**Comment by olojuwin** [comment]
> #19
> 
> 请提供一下启动命令和所有修改的内容。 同时麻烦截图一下卡死时候的nvitop界面，看一下卡住时候的显存和内存占用。卡死的一种可能是显存/内存溢出，请确保机器上的显存/内存足够。

显存是足够的，没有cuda 报错

**Comment by Peterande** [comment]
https://github.com/lyuwenyu/RT-DETR/issues/242

根据这个issue，应该是数据集存在一个batch全为无标签的情况，请尝试预处理数据集，避免这种情况发生。我会在有空的时候看一下能否在代码层面绕开这个消息。

**Comment by Peterande** [comment]
为了验证，可以尝试一下提高每个gpu的batchsize到和单卡一个水平，看会不会卡死，这能避免一个batch全是空的极端情况。因为看使用的是A100。80G显存应该绰绰有余

**Comment by leoxxxxxD** [comment]
我是8卡训练，在7epoch的固定一个iter卡死，没有日志输出，也不报错

**Comment by Peterande** [comment]
> [lyuwenyu/RT-DETR#242](https://github.com/lyuwenyu/RT-DETR/issues/242)
> 
> 根据这个issue，应该是数据集存在一个batch全为无标签的情况，请尝试预处理数据集，避免这种情况发生。我会在有空的时候看一下能否在代码层面绕开这种case。

是有这种一个batch全空的情况吗？

**Comment by yarkable** [comment]
我的case是，加了一些负样本去抑制误检，然后单卡正常运行，多卡必卡死。后面修改了一下 getitem，要是没有 box 的话就继续 random 一个 idx 取数据，直到拿到有 box 的图片。这样是 ok 了，但是负样本就没有起到作用了，怀疑跟 torchvision 的 transform 有关系

**Comment by cqlp1314** [comment]
> 我是8卡训练，在7epoch的固定一个iter卡死，没有日志输出，也不报错

我4卡训练,在55 epoch固定卡死,shuffle设为False也是固定55 epoch卡死

**Comment by ffxxy** [comment]
在train_one_epoch函数里面加:
# 修改bug
        #  ===============================================================================================================
        # 当前 rank 是否标签全为空
        local_empty = torch.tensor(
            int(all(len(t['boxes']) == 0 for t in targets)),
            device=samples.device  # 或 torch.device("cuda")，确保是 GPU 上的 tensor
        )

        # 将所有 rank 的 local_empty 相加（相当于全局汇总有几个 rank 是空的）
        dist.all_reduce(local_empty, op=dist.ReduceOp.SUM)

        # 只要有任何一个 rank 是空的（local_empty >= 1），所有 rank 就统一 continue
        if local_empty.item() > 0:
            if dist.get_rank() == 0:
                print(f"========================[Rank 0] Epoch {epoch}, iter {i}: 检测到至少一个 rank 的标签为空，全体跳过该 batch===========================")
            continue
        #  ===============================================================================================================
